File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/CucumberExpressionTokenizer.java
package io.cucumber.cucumberexpressions;

import io.cucumber.cucumberexpressions.Ast.Token;
import io.cucumber.cucumberexpressions.Ast.Token.Type;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.PrimitiveIterator.OfInt;

import static io.cucumber.cucumberexpressions.CucumberExpressionException.createCantEscape;
import static io.cucumber.cucumberexpressions.CucumberExpressionException.createTheEndOfLineCanNotBeEscaped;

final class CucumberExpressionTokenizer {

    List<Token> tokenize(String expression) {
        List<Token> tokens = new ArrayList<>();
        tokenizeImpl(expression).forEach(tokens::add);
        return tokens;
    }

    private Iterable<Token> tokenizeImpl(String expression) {
        return () -> new TokenIterator(expression);
    }

    private static class TokenIterator implements Iterator<Token> {

        private final String expression;
        private final OfInt codePoints;

        private StringBuilder buffer = new StringBuilder();
        private Type previousTokenType = null;
        private Type currentTokenType = Type.START_OF_LINE;
        private boolean treatAsText;
        private int bufferStartIndex;
        private int escaped;

        TokenIterator(String expression) {
            this.expression = expression;
            this.codePoints = expression.codePoints().iterator();
        }

        private Token convertBufferToToken(Type tokenType) {
            int escapeTokens = 0;
            if (tokenType == Type.TEXT) {
                escapeTokens = escaped;
                escaped = 0;
            }
            int consumedIndex = bufferStartIndex + buffer.codePointCount(0, buffer.length()) + escapeTokens;
            Token t = new Token(buffer.toString(), tokenType, bufferStartIndex, consumedIndex);
            buffer = new StringBuilder();
            this.bufferStartIndex = consumedIndex;
            return t;
        }

        private void advanceTokenTypes() {
            previousTokenType = currentTokenType;
            currentTokenType = null;
        }

        private Type tokenTypeOf(Integer token, boolean treatAsText) {
            if (!treatAsText) {
                return Token.typeOf(token);
            }
            if (Token.canEscape(token)) {
                return Type.TEXT;
            }
            throw createCantEscape(expression, bufferStartIndex + buffer.codePointCount(0, buffer.length()) + escaped);
        }

        private boolean shouldContinueTokenType(Type previousTokenType,
                Type currentTokenType) {
            return currentTokenType == previousTokenType
                    && (currentTokenType == Type.WHITE_SPACE || currentTokenType == Type.TEXT);
        }

        @Override
        public boolean hasNext() {
            return previousTokenType != Type.END_OF_LINE;
        }

        @Override
        public Token next() {
            if (!hasNext()) {
                throw new NoSuchElementException();
            }
            if (currentTokenType == Type.START_OF_LINE) {
                Token token = convertBufferToToken(currentTokenType);
                advanceTokenTypes();
                return token;
            }

            while (codePoints.hasNext()) {
                int codePoint = codePoints.nextInt();
                if (!treatAsText && Token.isEscapeCharacter(codePoint)) {
                    escaped++;
                    treatAsText = true;
                    continue;
                }
                currentTokenType = tokenTypeOf(codePoint, treatAsText);
                treatAsText = false;

                if (previousTokenType == Type.START_OF_LINE ||
                        shouldContinueTokenType(previousTokenType, currentTokenType)) {
                    advanceTokenTypes();
                    buffer.appendCodePoint(codePoint);
                } else {
                    Token t = convertBufferToToken(previousTokenType);
                    advanceTokenTypes();
                    buffer.appendCodePoint(codePoint);
                    return t;
                }
            }

            if (buffer.length() > 0) {
                Token token = convertBufferToToken(previousTokenType);
                advanceTokenTypes();
                return token;
            }

            currentTokenType = Type.END_OF_LINE;
            if (treatAsText) {
                throw createTheEndOfLineCanNotBeEscaped(expression);
            }
            Token token = convertBufferToToken(currentTokenType);
            advanceTokenTypes();
            return token;
        }

    }

}

File: cucumber-expressions/ruby/lib/cucumber/cucumber_expressions/cucumber_expression_tokenizer.rb
require 'cucumber/cucumber_expressions/ast'
require 'cucumber/cucumber_expressions/errors'

module Cucumber
  module CucumberExpressions
    class CucumberExpressionTokenizer
      def tokenize(expression)
        @expression = expression
        tokens = []
        @buffer = []
        previous_token_type = TokenType::StartOfLine
        treat_as_text = false
        @escaped = 0
        @buffer_start_index = 0

        codepoints = expression.codepoints

        if codepoints.empty?
          tokens.push(Token.new(TokenType::StartOfLine, '', 0, 0))
        end

        codepoints.each do |codepoint|
          if !treat_as_text && Token.isEscapeCharacter(codepoint)
            @escaped += 1
            treat_as_text = true
            next
          end
          current_token_type = token_type_of(codepoint, treat_as_text)
          treat_as_text = false

          if should_create_new_token(previous_token_type, current_token_type)
            token = convert_buffer_to_token(previous_token_type)
            previous_token_type = current_token_type
            @buffer.push(codepoint)
            tokens.push(token)
          else
            previous_token_type = current_token_type
            @buffer.push(codepoint)
          end
        end

        if @buffer.length > 0
          token = convert_buffer_to_token(previous_token_type)
          tokens.push(token)
        end

        if treat_as_text
          raise TheEndOfLineCannotBeEscaped.new(expression)
        end

        tokens.push(
            Token.new(TokenType::EndOfLine, '', codepoints.length, codepoints.length)
        )
        tokens
      end

      private

      # TODO: Make these lambdas

      def convert_buffer_to_token(token_type)
        escape_tokens = 0
        if token_type == TokenType::Text
          escape_tokens = @escaped
          @escaped = 0
        end

        consumed_index = @buffer_start_index + @buffer.length + escape_tokens
        t = Token.new(
            token_type,
            @buffer.map { |codepoint| codepoint.chr(Encoding::UTF_8) }.join(''),
            @buffer_start_index,
            consumed_index
        )
        @buffer = []
        @buffer_start_index = consumed_index
        t
      end

      def token_type_of(codepoint, treat_as_text)
        unless treat_as_text
          return Token.typeOf(codepoint)
        end
        if Token.canEscape(codepoint)
          return TokenType::Text
        end
        raise CantEscape.new(
            @expression,
            @buffer_start_index + @buffer.length + @escaped
        )
      end

      def should_create_new_token(previous_token_type, current_token_type)
        current_token_type != previous_token_type ||
            (current_token_type != TokenType::WhiteSpace && current_token_type != TokenType::Text)
      end
    end
  end
end
